sqlccl: use WriteBatch command in Restore implementation

danhhz opened the pull request:
This is performance work, no (intended) behavior change. The WriteBatch
command allows us to send a batch of Puts at once without all the
serializations & deserializations that would usually happen along the
way.

```
name              old time/op    new time/op     delta
ClusterRestore-8    14.1µs ±20%      7.4µs ±14%  -47.32%  (p=0.008 n=5+5)

name              old speed      new speed       delta
ClusterRestore-8  8.29MB/s ±23%  15.58MB/s ±13%  +87.93%  (p=0.008 n=5+5)

name              old alloc/op   new alloc/op    delta
ClusterRestore-8    19.2kB ± 2%      7.2kB ± 4%  -62.60%  (p=0.016 n=4+5)

name              old allocs/op  new allocs/op   delta
ClusterRestore-8      20.0 ±15%        3.0 ± 0%  -85.00%  (p=0.029 n=4+4)
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13586)
<!-- Reviewable:end -->

danhhz committed:
sqlccl: use WriteBatch command in Restore implementation

This is performance work, no (intended) behavior change. The WriteBatch
command allows us to send a batch of Puts at once without all the
serializations & deserializations that would usually happen along the
way.

Also adds a NewIterator to RocksDBSstFileReader since it's a better fit
for this code than the callback-based Iterate. (And handling the
tombstones for incremental restore, when we do it, is an even worse
fit for Iterate)

name              old time/op    new time/op     delta
ClusterRestore-8    15.3µs ±16%      5.8µs ±18%   -62.10%  (p=0.008 n=5+5)

name              old speed      new speed       delta
ClusterRestore-8  7.58MB/s ±15%  19.95MB/s ±16%  +163.10%  (p=0.008 n=5+5)

name              old alloc/op   new alloc/op    delta
ClusterRestore-8    19.3kB ± 2%      7.5kB ± 7%   -61.19%  (p=0.008 n=5+5)

name              old allocs/op  new allocs/op   delta
ClusterRestore-8      30.6 ±44%        2.2 ±33%   -92.65%  (p=0.000 n=5+4)

tamird commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-KcyLhgiKbxgiBVlnIoH:-KcyLhgiKbxgiBVlnIoI:b-nvolrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a3d354eb8047a58794921e07882e7fa0277f92/pkg/ccl/sqlccl/backup.go#L301)):*
> ```Go
> 
> 	var b engine.RocksDBBatchBuilder
> 	var batchStartKey []byte
> ```

this seems quite an odd pattern - why not pass these as arguments to `sendWriteBatch`?

these things are being mutated all over the place, and i don't see a good reason for it. can you help me understand?

---

*[pkg/ccl/sqlccl/backup.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-KcyLAeRpyk-soSUnEkd:-KcyLAeS6FIfojKbfLSC:bgy7mo) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a3d354eb8047a58794921e07882e7fa0277f92/pkg/ccl/sqlccl/backup.go#L320)):*
> ```Go
> 			wg.Done(nil)
> 		}(
> 			append([]byte(nil), batchStartKey...),
> ```

why the copies?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-KcyLhgiKbxgiBVlnIoH:-Kd1gWBrI6gPhvEs0juB:bpxfaus) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a3d354eb8047a58794921e07882e7fa0277f92/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this seems quite an odd pattern - why not pass these as arguments to `sendWriteBatch`?

these things are being mutated all over the place, and i don't see a good reason for it. can you help me understand?
</blockquote></details>

Good point. Extracted this all to a struct. Does that help?

(I don't love embedding this context, but it was pretty awkward to pass both it and the cancel func to send (via Put). Ideas?)

---

*[pkg/ccl/sqlccl/backup.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-KcyLAeRpyk-soSUnEkd:-Kd1gdBgm9XRrO4D_uJo:b-l4396t) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a3d354eb8047a58794921e07882e7fa0277f92/pkg/ccl/sqlccl/backup.go#L320)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why the copies?
</blockquote></details>

Added a comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Thanks for the reviews! RFAL

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-:-Kd1h2uAtZSTfugMZ__S:b-21u8wv)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-KcyLhgiKbxgiBVlnIoH:-Kd1j-AtS6KVaLpAYRas:b-m35tzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a3d354eb8047a58794921e07882e7fa0277f92/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Good point. Extracted this all to a struct. Does that help?

(I don't love embedding this context, but it was pretty awkward to pass both it and the cancel func to send (via Put). Ideas?)
</blockquote></details>

Heh, it looks quite a bit more convoluted now. For example, `(*writeBatcher).Put` internally copies its arguments because it knows its caller will overwrite them - that's just one example, but such implicit contracts make this code hard to read.

---

*[pkg/ccl/sqlccl/backup.go, line 323 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd1jMphRjnmxJabE2xc:-Kd1jMphRjnmxJabE2xd:b-e7ocwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f267cb96768cf19ad112f93c82fcc4e5bd144173/pkg/ccl/sqlccl/backup.go#L323)):*
> ```Go
> 
> 	wb.batchStartKey = wb.batchStartKey[:0]
> 	wb.batchEndKey = wb.batchEndKey[:0]
> ```

why do you do this here at all?

---

*[pkg/ccl/sqlccl/backup.go, line 374 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd1img12G2B1uwS1fZ8:-Kd1img12G2B1uwS1fZ9:b3au0dd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f267cb96768cf19ad112f93c82fcc4e5bd144173/pkg/ccl/sqlccl/backup.go#L374)):*
> ```Go
> 			// table(s) we're interested in.
> 			if log.V(3) {
> 				log.Infof(ctx, "Skipping %s %s\n", kv.Key.Key, v.PrettyPrint())
> ```

no need for the trailing newline, and please don't capitalize messages (throughout, apparently)

---

*[pkg/ccl/sqlccl/backup.go, line 387 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd1iuBNcWBC9IkV7nGq:-Kd1iuBNcWBC9IkV7nGr:b-y97260) ([raw file](https://github.com/cockroachdb/cockroach/blob/f267cb96768cf19ad112f93c82fcc4e5bd144173/pkg/ccl/sqlccl/backup.go#L387)):*
> ```Go
> 			log.Infof(ctx, "Put %s %s\n", kv.Key.Key, v.PrettyPrint())
> 		}
> 		b.Put(kv.Key, v.RawBytes)
> ```

why CPut -> Put?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-KcyLhgiKbxgiBVlnIoH:-Kd25G0UvVWakaWb3Jdq:b-rfn6hu) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a3d354eb8047a58794921e07882e7fa0277f92/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Heh, it looks quite a bit more convoluted now. For example, `(*writeBatcher).Put` internally copies its arguments because it knows its caller will overwrite them - that's just one example, but such implicit contracts make this code hard to read.
</blockquote></details>

Could you be a little more clear about whether you'd like better documentation of the contracts or whether you feel the code first needs structural changes, plz?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Okay, hopefully this clarified some stuff. RFAL

---

Review status: 2 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-KcyLhgiKbxgiBVlnIoH:-Kd72RKUQD6omLPjxoJp:b-j7p42l) ([raw file](https://github.com/cockroachdb/cockroach/blob/81a3d354eb8047a58794921e07882e7fa0277f92/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Could you be a little more clear about whether you'd like better documentation of the contracts or whether you feel the code first needs structural changes, plz?
</blockquote></details>

Now with less premature optimization!

---

*[pkg/ccl/sqlccl/backup.go, line 323 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd1jMphRjnmxJabE2xc:-Kd72ObGN3GA-4qVhXOy:bvfhtsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f267cb96768cf19ad112f93c82fcc4e5bd144173/pkg/ccl/sqlccl/backup.go#L323)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why do you do this here at all?
</blockquote></details>

removed this code

---

*[pkg/ccl/sqlccl/backup.go, line 374 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd1img12G2B1uwS1fZ8:-Kd7-JfOpplT6FWEIGEJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f267cb96768cf19ad112f93c82fcc4e5bd144173/pkg/ccl/sqlccl/backup.go#L374)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

no need for the trailing newline, and please don't capitalize messages (throughout, apparently)
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 387 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd1iuBNcWBC9IkV7nGq:-Kd7-dNSfv5WT922QUTF:b2i9su7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f267cb96768cf19ad112f93c82fcc4e5bd144173/pkg/ccl/sqlccl/backup.go#L387)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why CPut -> Put?
</blockquote></details>

Can't do a CPut in RocksDBBatchBuilder (and WriteBatch checks that the range is empty). Anyway, there was no particular reason it was a CPut before, so I don't feel we're losing anything

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-:-Kd72Y96tCp6HN6sHF29:bot11cb)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 295 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd78Ld1ZVl6hGL7yVJO:-Kd78Ld1ZVl6hGL7yVJP:b-4vhdw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L295)):*
> ```Go
> 	}
> 
> 	ctx, cancel := context.WithCancel(ctx)
> ```

`cancel` is idempotent, don't forget to `defer` it here.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd76mtdB6yPapU5EABc:-Kd76mtdB6yPapU5EABd:b7lnouc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L301)):*
> ```Go
> 
> 	var wg util.WaitGroupWithError
> 	var b engine.RocksDBBatchBuilder
> ```

one thing you could do here is put all these guys in a single struct since they're all reset together.

---

*[pkg/ccl/sqlccl/backup.go, line 307 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd7789pqSa363UWOIrS:-Kd7789pqSa363UWOIrT:b82vbpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L307)):*
> ```Go
> 		// The end key of the WriteBatch request is exclusive, but batchEndKey
> 		// is currently the largest key in the batch. Increment it.
> 		end := roachpb.Key(batchEndKey).PrefixEnd()
> ```

The comment above says you're incrementing, but this is doing more than increment. Why not .Next()?

---

*[pkg/ccl/sqlccl/backup.go, line 314 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd78X8Gbvt2csDsuOIY:-Kd78X8Gbvt2csDsuOIZ:bp4s101) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L314)):*
> ```Go
> 		wg.Add(1)
> 		func() {
> 			if err := db.WriteBatch(ctx, batchStartKey, end, b.Finish()); err != nil {
> ```

this code is meant to be async, I think, but it is problematic that it closes over `batchStartKey` and `b` since they are immediately reset after this goroutine is started.

---

*[pkg/ccl/sqlccl/backup.go, line 357 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd77Rnc7aNElXtTkP61:-Kd77Rnc7aNElXtTkP62:bfb3320) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L357)):*
> ```Go
> 		}
> 
> 		iter := sst.NewIterator(false)
> ```

what's the difference between this style and the previous use of `sst.Iterate`? The only difference I can spot is the need to explicitly manage the keys

---

*[pkg/ccl/sqlccl/backup.go, line 383 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd77xJcssWapAv9A84X:-Kd77xJdsE6mzC5u8q1M:bpwkcwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L383)):*
> ```Go
> 
> 			if log.V(3) {
> 				log.Infof(ctx, "put %s %s", key.Key, value.PrettyPrint())
> ```

nit: elsewhere in sql we use log lines of the form `"CPut %s -> %s"`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup.go, line 295 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd78Ld1ZVl6hGL7yVJO:-Kd7B1I5r90C3YEJ2jdd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L295)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`cancel` is idempotent, don't forget to `defer` it here.
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd76mtdB6yPapU5EABc:-Kd7C4k37QJQjvRze8V8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

one thing you could do here is put all these guys in a single struct since they're all reset together.
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 307 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd7789pqSa363UWOIrS:-Kd7C5wtnEiwr-SIBdGT:bu4jz2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L307)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

The comment above says you're incrementing, but this is doing more than increment. Why not .Next()?
</blockquote></details>

Didn't know about Next. Done

---

*[pkg/ccl/sqlccl/backup.go, line 314 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd78X8Gbvt2csDsuOIY:-Kd7C7ubn0iovFdaaW0l:b-xx61yf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L314)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this code is meant to be async, I think, but it is problematic that it closes over `batchStartKey` and `b` since they are immediately reset after this goroutine is started.
</blockquote></details>

Whoops

---

*[pkg/ccl/sqlccl/backup.go, line 357 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd77Rnc7aNElXtTkP61:-Kd7Cb59ZtLVDBO0rswc:b-h0u2xy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L357)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's the difference between this style and the previous use of `sst.Iterate`? The only difference I can spot is the need to explicitly manage the keys
</blockquote></details>

I find this easier to read and it'll be necessary when we handle the tombstoning for incremental restore

---

*[pkg/ccl/sqlccl/backup.go, line 383 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd77xJcssWapAv9A84X:-Kd7CJlw48tbIlwk3L0G:bquwmes) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L383)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: elsewhere in sql we use log lines of the form `"CPut %s -> %s"`
</blockquote></details>

Sure, though it really is just a Put now. If you add the same key twice, it'll happily overwrite

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-:-Kd7G9tJN3IyILbUq2ti:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd76mtdB6yPapU5EABc:-Kd7K54gf6Uk0T8LwDrF:b-ef57t4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

if you name the type you can `b := theType{}`. optional.

---

*[pkg/ccl/sqlccl/backup.go, line 383 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd77xJcssWapAv9A84X:-Kd7KTzl1kB--aP6fXkj:b-6w6887) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L383)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Sure, though it really is just a Put now. If you add the same key twice, it'll happily overwrite
</blockquote></details>

Yeah, didn't mean to call out the specific operation, just the style of log line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd76mtdB6yPapU5EABc:-Kd7KbzxCOf6jN5hSNwU:bg9q5ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

if you name the type you can `b := theType{}`. optional.
</blockquote></details>

`b = theType{}` I mean. at the bottom of `sendWriteBatch`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Thanks for the reviews!!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 301 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd76mtdB6yPapU5EABc:-Kd7NcmksbTAZWWxkiq9:b-866npd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L301)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`b = theType{}` I mean. at the bottom of `sendWriteBatch`.
</blockquote></details>

ack. I think I'm going to leave this as-is

---

*[pkg/ccl/sqlccl/backup.go, line 383 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-Kd77xJcssWapAv9A84X:-Kd7NkJfuoXTI8ZZqORg:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ccafb31afeaa6a84c62e3ce175c8c0811d64401/pkg/ccl/sqlccl/backup.go#L383)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Yeah, didn't mean to call out the specific operation, just the style of log line.
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13586#-:-Kd7Np9HppWUJ4fJpxeL:bmvd9tv)*
<!-- Sent from Reviewable.io -->
